tests/v2/validation/snapshot/snapshot_restore_k8s_upgrade_test.go

```diff
@@ -0,0 +1,77 @@
+//go:build (validation || extended) && !infra.any && !infra.aks && !infra.eks && !infra.gke && !infra.rke2k3s && !cluster.any && !cluster.custom && !cluster.nodedriver && !sanity && !stress
+
+package snapshot
+
+import (
+	"testing"
+
+	"github.com/rancher/rancher/tests/framework/clients/rancher"
+	"github.com/rancher/rancher/tests/framework/extensions/etcdsnapshot"
+
+	"github.com/rancher/rancher/tests/framework/pkg/config"
+	"github.com/rancher/rancher/tests/framework/pkg/session"
+
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+type SnapshotRestoreK8sUpgradeTestSuite struct {
+	suite.Suite
+	session        *session.Session
+	client         *rancher.Client
+	clustersConfig *etcdsnapshot.Config
+}
+
+func (s *SnapshotRestoreK8sUpgradeTestSuite) TearDownSuite() {
+	s.session.Cleanup()
+}
+
+func (s *SnapshotRestoreK8sUpgradeTestSuite) SetupSuite() {
+	testSession := session.NewSession()
+	s.session = testSession
+
+	s.clustersConfig = new(etcdsnapshot.Config)
+	config.LoadConfig(etcdsnapshot.ConfigurationFileKey, s.clustersConfig)
+
+	client, err := rancher.NewClient("", testSession)
+	require.NoError(s.T(), err)
+
+	s.client = client
+}
+
+func (s *SnapshotRestoreK8sUpgradeTestSuite) TestSnapshotRestoreK8sUpgrade() {
+	snapshotRestoreK8sVersion := &etcdsnapshot.Config{
+		UpgradeKubernetesVersion: s.clustersConfig.UpgradeKubernetesVersion,
+		SnapshotRestore:          "kubernetesVersion",
+	}
+
+	snapshotRestoreAll := &etcdsnapshot.Config{
+		UpgradeKubernetesVersion: s.clustersConfig.UpgradeKubernetesVersion,
+		SnapshotRestore:          "all",
+	}
+
+	tests := []struct {
+		name         string
+		etcdSnapshot *etcdsnapshot.Config
+		client       *rancher.Client
+	}{
+		{"Restore Kubernetes version and etcd", snapshotRestoreK8sVersion, s.client},
+		{"Restore cluster config, Kubernetes version and etcd", snapshotRestoreAll, s.client},
+	}
+
+	for _, tt := range tests {
+		s.Run(tt.name, func() {
+			snapshotRestore(s.T(), s.client, s.client.RancherConfig.ClusterName, tt.etcdSnapshot, false)
+		})
+	}
+}
+
+func (s *SnapshotRestoreK8sUpgradeTestSuite) TestSnapshotRestoreK8sUpgradeDynamicInput() {
+	snapshotRestore(s.T(), s.client, s.client.RancherConfig.ClusterName, s.clustersConfig, false)
+}
+
+// In order for 'go test' to run this suite, we need to create
+// a normal test function and pass our suite to suite.Run
+func TestSnapshotRestoreK8sUpgradeTestSuite(t *testing.T) {
+	suite.Run(t, new(SnapshotRestoreK8sUpgradeTestSuite))
```

Comment:
since upgradeStrategy requires a k8s upgrade, should that just be a subset of tests in this test suite? or would we want to keep them separate for some reason?

Reply:
I would want to keep them separate personally. I think it's fine leaving them as is